packages/components/tooltip/src/tooltip.vue

```diff
@@ -68,162 +67,173 @@ import {
-        contentRef.value?.contentRef?.popperContentRef
-      return popperContent && popperContent.contains(document.activeElement)
-    }
+const id = useId()
+// TODO any is temporary, replace with `InstanceType<typeof ElPopper> | null` later
+const popperRef = ref<any>(null)
+// TODO any is temporary, replace with `InstanceType<typeof ElTooltipContent> | null` later
+const contentRef = ref<any>(null)
+
+const updatePopper = () => {
+  const popperComponent = unref(popperRef)
+  if (popperComponent) {
+    popperComponent.popperInstanceRef?.update()
+  }
+}
+const open = ref(false)
+const toggleReason = ref<Event | undefined>(undefined)
 
-    onDeactivated(() => open.value && hide())
-
-    return {
-      compatShowAfter,
-      compatShowArrow,
-      popperRef,
-      contentRef,
-      open,
-      hide,
-      isFocusInsideContent,
-      updatePopper,
-      onOpen,
-      onClose,
+const { show, hide, hasUpdateHandler } = useModelToggle({
+  indicator: open,
+  toggleReason,
+})
+
+const { onOpen, onClose } = useDelayedToggle({
+  showAfter: compatShowAfter,
+  hideAfter: toRef(props, 'hideAfter'),
+  open: show,
+  close: hide,
+})
+
+const controlled = computed(
+  () => isBoolean(props.visible) && !hasUpdateHandler.value
+)
+
+provide(TOOLTIP_INJECTION_KEY, {
+  controlled,
+  id,
+  open: readonly(open),
+  trigger: toRef(props, 'trigger'),
+  onOpen: (event?: Event) => {
+    onOpen(event)
+  },
+  onClose: (event?: Event) => {
+    onClose(event)
+  },
+  onToggle: (event?: Event) => {
+    if (unref(open)) {
+      onClose(event)
+    } else {
+      onOpen(event)
     }
   },
+  onShow: () => {
+    emit('show', toggleReason.value)
+  },
+  onHide: () => {
+    emit('hide', toggleReason.value)
+  },
+  onBeforeShow: () => {
+    emit('before-show', toggleReason.value)
+  },
+  onBeforeHide: () => {
+    emit('before-hide', toggleReason.value)
+  },
+  updatePopper,
+})
+
+watch(
+  () => props.disabled,
+  (disabled) => {
+    if (disabled && open.value) {
+      open.value = false
+    }
+  }
+)
+
+const isFocusInsideContent = () => {
+  const popperContent: HTMLElement | undefined =
+    contentRef.value?.contentRef?.popperContentRef
+  return popperContent && popperContent.contains(document.activeElement)
+}
+
+onDeactivated(() => open.value && hide())
+
+defineExpose({
+  /**
+   * @description el-popper component instance
+   */
+  popperRef,
+  /**
+   * @description el-tooltip-content component instance
+   */
+  contentRef,
+  /**
+   * @description validate current focus event is trigger inside el-tooltip-content
+   */
+  isFocusInsideContent,
+  /**
+   * @description update el-popper component instance
+   */
+  updatePopper,
+  /**
+   * @description expose onOpen function to mange el-tooltip open state
+   */
+  onOpen,
+  /**
+   * @description expose onOpen function to mange el-tooltip open state
+   */
+  onClose,
+  /**
+   * @description expose disable state
+   */
+  disabled: props.disabled,
```

Comment:
No need to expose it.

Reply:
I think we could remove it as it is only used in the test case here.

https://github.com/element-plus/element-plus/blob/2725dca47a8715d9b3a4d95df57e528f6e7da4c1/packages/components/slider/__tests__/slider.test.tsx#L75-L77

Also, in this test case we need to make some changes:

```diff
- const slider = wrapper.vm.$refs.slider 
- const tooltip = slider.$refs.firstButton.$refs.tooltip 
+ const tooltip = wrapper.findComponent({ name: 'ElTooltip' }).vm
```